packages/core/src/app/operator/block-db.ts

```diff
@@ -0,0 +1,146 @@
+  BLOCK: new BaseKey('b', ['uint32']),
+}
+
+/**
+ * Simple BlockDB implementation.
+ */
+export class DefaultBlockDB implements BlockDB {
+  private vars: KeyValueStore
+  private blocks: KeyValueStore
+
+  /**
+   * Initializes the database wrapper.
+   * @param db Database to store values in.
+   */
+  constructor(private db: KeyValueStore) {
+    this.vars = this.db.bucket(PREFIXES.VARS)
+    this.blocks = this.db.bucket(PREFIXES.BLOCKS)
+  }
+
+  /**
+   * @returns the next plasma block number.
+   */
+  public async getNextBlockNumber(): Promise<number> {
+    // NOTE: You could theoretically cache the next block number as to avoid
+    // having to read from the database repeatedly. However, this introduces
+    // the need to ensure that the cached value and stored value never fall out
+    // of sync. It's probably better to hold off on implementing caching until
+    // this becomes a performance bottleneck.
+
+    const buf = await this.vars.get(KEYS.NEXT_BLOCK)
+    return buf.readUInt32BE(0)
+  }
+
+  /**
+   * Adds a state update to the list of updates to be published in the next
+   * plasma block.
+   * @param stateUpdate State update to publish in the next block.
+   * @returns a promise that resolves once the update has been added.
+   */
+  public async addPendingStateUpdate(stateUpdate: StateUpdate): Promise<void> {
+    const block = await this.getNextBlockStore()
+    const start = stateUpdate.id.start
+    const end = stateUpdate.id.end
+
+    // TODO: Figure out how to implement locking here so two state updates
+    // can't be added at the same time.
+
+    if (await block.has(start, end)) {
+      throw new Error('Block already contains a state update over that range.')
+    }
+
+    const value = Buffer.from(JSON.stringify(stateUpdate))
+    await block.put(start, end, value)
+  }
+
+  /**
+   * @returns the list of state updates waiting to be published in the next
+   * plasma block.
+   */
+  public async getPendingStateUpdates(): Promise<StateUpdate[]> {
+    const blockNumber = await this.getNextBlockNumber()
+    return this.getStateUpdates(blockNumber)
+  }
+
+  /**
+   * Computes the Merkle Interval Tree root of a given block.
+   * @param blockNumber Block to compute a root for.
+   * @returns the root of the block.
+   */
+  public async getMerkleRoot(blockNumber: number): Promise<string> {
+    const stateUpdates = await this.getStateUpdates(blockNumber)
+
+    const leaves = stateUpdates.map((stateUpdate) => {
+      const encodedStateUpdate = JSON.stringify(stateUpdate) // TODO: Actually encode this.
+      return {
+        end: stateUpdate.id.end,
+        data: encodedStateUpdate,
+      }
+    })
+    const tree = new MerkleIntervalTree({
+      leaves,
+    })
+    return tree.root
+  }
+
+  /**
+   * Finalizes the next plasma block so that it can be published.
+   * @returns a promise that resolves once the block has been finalized.
+   */
+  public async finalizeNextBlock(): Promise<void> {
```

Comment:
Not sure that it matters, but there's a race condition here if called concurrently.

Reply:
Yeah I could definitely use some input on managing concurrency. Seems like we might have to manage it on a global/OS level because `leveldb` isn't safe between different processes.

Reply:
Specifically important for the operator because accidentally bumping the block number twice would break a lot of things.

Reply:
We'll have to look into this more. It could be as simple as using a lib that wraps some simple single-threaded bool-flipping logic. We should be able to use JS's single-threaded nature to enforce order where necessary on multiple awaited calls.